applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_strategy.h

```diff
@@ -461,61 +468,210 @@ class MPMStrategy
     void Initialize() override
     {
     }
 
     /**
-    the problem of interest is solved
+     * @brief Performs all the required operations that should be done (for each step) before solving the solution step.
      */
-    double Solve() override
+    void InitializeSolutionStep() override
     {
-        // Check which nodes and elements are ACTIVE and populate the MPM model part
-        KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - Start" <<std::endl;
-        KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Search Element - Start" <<std::endl;
-        this->SearchElement(mr_grid_model_part, mr_mpm_model_part);
-
-        // Only perform this once
+        // Initialize solving strategy: only to be done at the beginning of time step
         mp_solving_strategy->Initialize();
 
         // The nodal initial conditions are computed
         KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - InitializeSolutionStep" <<std::endl;
         mp_solving_strategy->InitializeSolutionStep();
+    }
 
+    /**
+     * @brief Operation to predict the solution
+     */
+    void Predict() override
+    {
         KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - Predict" <<std::endl;
         mp_solving_strategy->Predict();
+    }
 
+    /**
+     * @brief Solves the current step. This function returns true if a solution has been found, false otherwise.
+     */
+    bool SolveSolutionStep() override
+    {
         // Do solution iterations
         KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - SolveSolutionStep" <<std::endl;
         mp_solving_strategy->SolveSolutionStep();
 
+        return true;
+    }
+
+    /**
+     * @brief Performs all the required operations that should be done (for each step) after solving the solution step.
+     */
+    void FinalizeSolutionStep() override
+    {
         // The nodal solution are mapped from mesh to MP
         KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - FinalizeSolutionStep" <<std::endl;
         mp_solving_strategy->FinalizeSolutionStep();
+    }
 
+    /**
+     * @brief Clears the internal storage
+     */
+    void Clear() override
+    {
         KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - Clear" <<std::endl;
         mp_solving_strategy->Clear();
+    }
+
+    /**
+     * @brief Search element connectivity for each particle
+     * @details A search is performed to know in which grid element the material point falls.
+     * If one or more material points fall in the grid element, the grid element is
+     * set to be active and its connectivity is associated to the material point
+     * element.
+     * STEPS:
+     * 1) All the elements are set to be INACTIVE
+     * 2) A searching is performed and the grid elements which contain at least a MP are set to be ACTIVE
+     *
+    */
+    virtual void SearchElement(
```

Comment:
you just moved this function right?

Btw this could be a separate fct to keep things modular (and for testing it separately)

it does not seem to be strictly related to the strategy

Reply:
yes I just moved the function above. Yes, indeed, this can be used, let say to know the background position of the FEM node inside an MPM background too. I can implement this outside the strategy. But I will do it when necessary.